### ISSUE TITLE
filename too long or special characters ONLY on ipad and iphone

### ISSUE THREAD

**Issue by mrcnee93** [rockstor/rockstor-core#2649]
i'm running rockstor on an old pc and it's working perfectly except for this strange behavior.

i can copy, paste, save file etc... from my macbook to the rockstor server but only save new file from my ipad and my iphone.
if i try to copy paste something from my mobile devices it shows an error "name too long etc..."

if there's some log file or anything i can share with you please guide me because i'm pretty new here

thanks

**Comment by phillxnet** [comment]
@mrcnee93 Thanks for the report. As there is too little to go on here for a GitHub issue just yet. Could you please start a new thread on our forum  here:
https://forum.rockstor.com/

GitHub issues really need an exact reproducer: otherwise we have no way to prove a fix. I.e. I did this and it broke so developer can do identical steps and see what went wrong. Then fix for that reproducer. So we try to use the forum to flesh out more details so that the GitHub issues are not super long and hard to follow.

All current developers are also forum members plus there will be folks there who can hopefully help with fleshing out the exact cause of this.

E.g:

> if i try to copy paste something from my mobile devices it shows an error "name too long etc..."

would need way more details - like all of them. File name length the platform version you are experiencing this with i.e. iOS version, how you are copying etc. So the forum is great for get all those details out to provide an exact reproducer (ideally) to write the issue title and content around. Not always possible but I think in this case it should be.

Have a chat on the forum to see if you can put in more details etc. And see if folks there can weed out what's going on here. We basically don't have limitations of filename lengths for example, beyond those of btrfs which are not normally reached.

Thanks for the report however - and lets hope the forum can wittle down what the problem you are seeing is.

And also we need non abbreviated error messages. You can always change filenames etc but keep the actual lengths and structure the same.

Glad it's working ok for you otherwise. Do be clear, in the forum post, on the exact version of Rockstor you are experiencing this with also.

Incidentally the special characters bit in the title is a hint of sorts. So watch for them/those. And report in the forum what triggers you observed failure.

By the way, our following doc section has some info on our main log file:
https://rockstor.com/docs/contribute/contribute.html#terminal-hint
and how to enable debugging to help with more info logged to that file.

Hope that helpls.

**Comment by Hooverdan96** [comment]
@mrcnee93 while you contemplating using the forum, I found a recent post on the apple discussions, where at least 13 people had the same issue. Unfortunately, nobody responded with a further explanation at this time.
https://discussions.apple.com/thread/254732969
Main difference was that this happened on a Samba setup on Ubuntu. So, it __might__ be more related to a iphone/pad - Samba interaction, rather than something caused by Rockstor.
All other references I found to this problem so far, were fairly old (from an iOS version standpoint) so I suspect none of those still apply.
But, I agree with @phillxnet, you have better chances finding someone on the forum that has gone through this problem ... but it's certainly no guarantee.

**Comment by Hooverdan96** [comment]
I was able to reproduce this on my own environment (in my case, it would put a 0KB file with the name on the share after triggering the "file name too long" error message. I think I found the solution to this problem. This is assuming, we don't have to worry about how Rockstor deals with a Samba share that's earmarked for Time Machine.

Based on this thread:

https://wiki.samba.org/index.php/Configure_Samba_to_Work_Better_with_Mac_OS_X

I've added the following settings to the Samba profile in the WebUI (which represents the `[Global]` section in the `/etc/samba/smb.conf` file:
```
min protocol = SMB2
vfs objects = fruit streams_xattr  
fruit:metadata = stream
fruit:model = MacSamba
fruit:posix_rename = yes 
fruit:veto_appledouble = no
fruit:nfs_aces = no
fruit:wipe_intentionally_left_blank_rfork = yes 
fruit:delete_empty_adfiles = yes 
```
Hit the submit button, and voila, the transfer from an iOS device (latest iOS 16.x version) - I had to `eject` in the Files.app and reconnect, 

Note: I have some other settings specific to my setup, which I'm not listing here, but the addition of the above made the difference.
I have not experimented with removing the min version of SMB2, but if I get around to it, and it works, then that would not be necessary ... (I read somewhere that Apple's smb client implementation runs on smb3 protocol but still does some thing using smb2, not sure what to make of that or whether it's even true).

@phillxnet, @FroggyFlox if @mrcnee93 can confirm that this fix works for them, I could take this information and put a thread on the forum, since that seems to be good information for iOS users

**Comment by FroggyFlox** [comment]
Awesome, @Hooverdan96 !

These smb settings look really familiar... could you check how much they differ with those applied when the user checks the "Time Machine" compatibility checkbox when creating a Samba export? My apologies if you have already tried that but I wonder if these settings themselves would be enough to fix it.

**Comment by Hooverdan96** [comment]
I just set up a share with the "Time Machine" flag added. Main differrence is, that the Apple compatibility entires are at the share level, whereas my test above set it at the Global level (applying to all shares).
Good part is from what I understand is, that any "conflicting" entries at the share level are overwriting the global settings, so it won't have an impact on the TimeMachine option/functionality.

Here are the settings that are added at the share level (when the TimeMachine flag is set in the WebUI):
```
    vfs objects = catia fruit streams_xattr
    fruit:timemachine = yes
    fruit:metadata = stream
    fruit:veto_appledouble = no
    fruit:posix_rename = no
    fruit:wipe_intentionally_left_blank_rfork = yes
    fruit:delete_empty_adfiles = yes
    fruit:encoding = private
    fruit:locking = none
    fruit:resource = file
```
they are not quite the same as from the Samba Wiki I referenced above.
the `catia` vfs object is not loaded
the `fruit:timemachine = yes` should be at the share level as is also pointed out in the Wiki
the `fruit:posix_rename` is different - the default if not specified is actually `yes` - to me it makes more sense to have that as `yes` but somebody can possibly explain to me why it should be `no`
the `fruit:locking = none` is actually the default setting, so I would assume it doesn't need to be set explicitly other than if somebody mucked with that in the Global settings
the `fruit:encoding = private` also is the default value for that setting
the `fruit:resource = file` also the default value
on the flip side (the wiki settings)
`fruit:metadata = stream` is the default value for that setting, which also explains why the vfs_objects order is important. In this case this setting means to pass on the stream to the next module in the vfs_objects stack, which in this case would be the `streams_xattr`.
`fruit:model = MacSamba` is not functionally relevant, but I guess nice to have the icon assigned in the Finder.app on apple machines

So, fundamentally, they are very similar a couple of exceptions.
`The vfs_objects = catia`  is truly only necessary when one backs up catia (CAD files) apparently. I don't think it hurts to have that in there, but that is not specifically affecting Apple machines, the same issue would occur when trying to copy "illegal" characters from a windows client. So having it, again, probably a good idea, but not directly connected to the Apple topic.

This one is not there in the TimeMachine settings:
`fruit:nfs_aces = no` which, according to the Wiki 
`will prevent macOS clients from modifying the UNIX mode of directories using NFS ACEs. An Access Control Entry (ACE) is part of the Access Control List (ACL).` I don't understand this enough, too me it seems this should be there to prevent changes to the way Linux handles directory access, but I might be totally wrong.

So, if one wants to have these settings at the global level, other than the `fruit=timemachine` setting could be moved up. I still think the `posix_rename` and `nfs_aces` settings are worth a discussion, unless someone has already determined why it's necessary for the timemachine ...

**Comment by Hooverdan96** [comment]
On the `fruit:nfs_aces` setting. Upon further reading, it seems that especially in the case where samba is used with Windows clients as well as Mac type clients, it's better to set this parameter to `no` to allow for maximum compatibility. If the share is for TimeMachine only, it won't matter, and could be even set to yes, but if the above settings are applied at the global level (for all samba exports) then it makes more sense to force a `no`.

One of the links I explored:
https://stackoverflow.com/questions/58496277/samba-4-10-server-config-using-module-vfs-fruit-changes-file-creation-mask-fo

**Comment by Hooverdan96** [comment]
I took the liberty to summarize my findings on the forum here:
https://forum.rockstor.com/t/configure-samba-to-work-better-with-apple-devices/8956

**Comment by Hooverdan96** [comment]
If we consider the summary on the rockstor forum (above link) the solution to this problem, then we can probably close this issue. Unless @mrcnee93 has some follow-up observations on this topic.

**Comment by phillxnet** [comment]
@Hooverdan96 Thanks for your excellent input/investigations here. As always much appreciated.

**Comment by 404ErrorUsrNtFnd** [comment]
Hi guys! I created an account specifically to express my gratitude to you. The information from this discussion helped me a lot! All the best to you!

**Comment by Hooverdan96** [comment]
@phillxnet I think this issue can now be closed, since a workaround is documented on the forum as well.

**Comment by phillxnet** [comment]
Closing as per @Hooverdan96's advise.
